azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,122 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import threading
+
+import traceback
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.future import ustr, PriorityQueue
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+
+
+def get_telemetry_service_handler(protocol_util):
+    return TelemetryServiceHandler(protocol_util)
+
+
+class TelemetryServiceHandler(ThreadHandlerInterface):
+    """
+    This Handler takes care of sending all telemetry out of the agent to Wireserver. It sends out data as soon as
+    there's any data available in the queue to send.
+    """
+
+    _THREAD_NAME = "TelemetryServiceHandler"
+    _MAX_TIMEOUT = datetime.timedelta(minutes=5).seconds
+
+    def __init__(self, protocol_util):
+        self._protocol = protocol_util.get_protocol()
+        self.should_run = True
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = PriorityQueue()
```

Comment:
This should be a FIFO queue rather than a priority queue, right? we should preserver the order in which the events are created when pushing them to the wireserver.

Reply:
I made it a priority queue mainly to always prioritize agent events over extension events. When the extension telemetry pipeline is in use properly, each extension can send out upto 300 events per run. Given that each call to telemetry can take almost ~3secs and that we send approx ~25-35events per request, the thread can be blocked for quite a while if we have a lot of events. I figured it would be useful to always give ourselves the priority of using our own pipeline and keep sending the other requests on the side

Reply:
I would use a regular queue until the priority is defined. At this point the items would be sorted and since they do not have a priority who knows in what order they will be sorted and sent.

Reply:
When the priority is defined, seems like you will need two sets of priorities, with the set for the agent taking priority over the set for extensions. However, the priority of events within each of those sets should respect the order in which the events were generated,

Reply:
> since they do not have a priority

just noticed you defined an order... let me check that

Reply:
seems like it only differentiates between agent/other events... you also need to respect the timestamp

Reply:
>However, the priority of events within each of those sets should respect the order in which the events were generated,

I agree about that. I was thinking of adding it but then figured it would be too much for this already big PR. I didn't do it mainly because we anyways fill the buffer up to the brim before sending out the requests and we group each request based on the ProviderID (the table it will end up in Kusto) so I figured for now this is good enough. I can add another timestamp parameter to it if you feel its a big deal though.

Reply:
also, this priority stuff may not send extensions events if the agent goes crazy and starts sending tons of events

Reply:
>I can add another timestamp parameter to it if you feel its a big deal though.

I think right now the code has the potential to never send old events if new ones keep coming up

Reply:
That is a possibility agreed, but that would be a pain even if we have a FIFO queue (like whenever the monitor thread adds the events, we wont process anything but the agent events). Since right now we start sending as soon as we enqueue anything in the queue, I think we might be able to balance it out. 

The converse is also true though, if we get a very chatty extension our telemetry would be blocked. The best case I guess for such a scenario would be to have separate telemetry calls for both the threads separately (something like the current code) but that has its own problems I suppose.

Reply:
Couldn't you simply modify your ordering to check both for `priority` and `time_generated` instead of just `priority` and that would solve the issue?

Reply:
After the last sync, we decided to remove the PriorityQueue altogether for now and just have a normal Queue approach. Incase later we feel that extension events are causing a lot of contention for Agent events and are blocking us heavily, we will add the PQ back with a priority + time_generated approach

Reply:
Done